File: Heap-Sort/js/leetcode_lc_40.js
const arr = [3,2,1]
const k = 2

/**
 * @param {number[]} arr
 * @param {number} k
 * @return {number[]}
 */
var getLeastNumbers = function(arr, k) {
    for(let i = Math.floor(arr.length/2) - 1 ; i >=0 ; i--){
        heapify(arr,i,arr.length)
    }
    const res = []
    for(let i = arr.length - 1 ; i >= 0; i--,k--){
        
        if(k === 0){ 
            return res
        }
        res.push(arr[0])

        const tmp = arr[i]
        arr[i] = arr[0]
        arr[0] = tmp

        heapify(arr,0,i)
    }
    console.log(arr)
    return res
};

var heapify = (arr,idx,len) => {
    const left = 2*idx + 1
    const right = 2*idx + 2
    let parent = idx

    if(left < len && arr[left] < arr[parent]){
        parent = left
    }

    if(right < len && arr[right] < arr[parent]){
        parent = right
    }

    if(parent !== idx){
        const tmp = arr[idx]
        arr[idx] = arr[parent]
        arr[parent] = tmp
        heapify(arr,parent,len)
    }
}

console.log(`${arr}数组中，最小的 ${k} 个数:`,getLeastNumbers(arr,k))

File: Heap-Sort/python/leetcode_lc_40.py
arr = [3,2,1]
k = 2

class Solution:
    def getLeastNumbers(self, arr, k):
        for i in range(int(len(arr)/2)-1,-1,-1):
            self.heapify(arr,i,len(arr))

        res = []

        for i in range(len(arr)-1,-1,-1):

            if k==0 :
                return res

            res.append(arr[0])
            arr[0],arr[i] = arr[i],arr[0]
            self.heapify(arr,0,i)
            k -= 1

        return res


    def heapify(self,arr,idx,n):
        left = 2*idx + 1
        right = 2*idx + 2
        parent = idx

        if left < n and arr[left] < arr[parent]:
            parent = left

        if right < n and arr[right] < arr[parent]:
            parent = right

        if parent != idx:
            arr[idx],arr[parent] = arr[parent],arr[idx]
            self.heapify(arr,parent,n)


solution = Solution()

print("数组{} 最小的 {} 个数: {}".format(arr,k,solution.getLeastNumbers(arr,k)))